mealpy.py

```diff
@@ -27,31 +29,35 @@
 
 KEYRING_SERVICENAME = BASE_DOMAIN
 
-# TODO(GH-4): Replace this feature toggle with cli option
-USE_KEYRING = False
-
 
 def load_config():
     schema = strictyaml.Map({
         'email_address': strictyaml.Email(),
+        'use_keyring': strictyaml.Bool()
     })
-
     root_dir = path.abspath(path.dirname(__file__))
-    with open(path.join(root_dir, 'config.yaml')) as config_file:
+    fname = path.join(root_dir, 'config.yaml')
+    # Create new file using template if not already existing
+    if not path.isfile(fname):
+        template = path.join(root_dir, 'config.template.yaml')
+        copyfile(template, fname)
+    with open(fname) as config_file:
         return strictyaml.load(config_file.read(), schema).data
 
+CONFIG = load_config()
 
-class MealPal():
 
-    def __init__(self, user, password=None):
+class MealPal:
+
+    def __init__(self, user, password):
         self.cookies = None
         self.user = user
         self.password = password
 
     def login(self):
         data = {
             'username': self.user,
-            'password': self.password or keyring.get_password(KEYRING_SERVICENAME, self.user),
+            'password': self.password,
```

Comment:
How do you avoid stale credentials for a long lived scheduled task? For adhoc user input, it's irrelevant, they need to restart the process to reconfigure. But for keyring users, the previous code was querying every time, which allows user to update the keyring entry and the next scheduled task would pick it up.

Otherwise, there's no value in apscheduler (maybe can close #8 as won't do), setting up a cronjob will get the job done.

Reply:
Oh, I see what you mean now. I was thinking that users probably won't change their passwords frequently, and was only handling password resets in case they entered their passwords wrongly at first or wanted to remove their passwords from keyring.

Reply:
Yes, this is an edge case. But in terms of best-practice, we're inadvertently caching the value and duplicating source-of-truth. That's why the original design was shuttling around `None` as a sentinel value. And deferring retrieving the value until it's actually required.

Reply:
Ok, I see your point. The reason I modified it initially was that `MealPal` was looking into `keyring` directly which was weird to me and I don't think it should be the `MealPal` class that should be the one handling this.

Reply:
If you want to keep `MealPal` class agnostic to credential management, both username and password should be passed into the `login` function. Keeping them as instance variables means the `MealPal` class owns the responsibility and, therefore, shoudl also be allowed to access `keyring` (a function call to get password, instead of variable access).

This might be okay class design. But then `MealPal` wouldn't be able to relogin on its own. To do so, maybe the entire `login` callback should be provided during instantiation.

Imo, maybe too much indirection and over-engineering this early on, the class is bound to continue to change a lot over the short term with the features we have planned.

Reply:
I think `MealPal` shouldn't have to know where the `user` and `password` strings come from, only that it receives these strings to be used as credentials, so it shouldn't know that `keyring` even exists. Instead the responsibility lies in the script running and creating the `MealPal` objects to fetch the right user and password from wherever and send them to the `MealPal` object, so the script should be the one handling this problem. 

So I think what can be changed is:
- remove `user` and `password` fields in `MealPal` and pass them into `login` as before.
- in `execute_reserve_meal` (the scheduled task), always opt to read the password from the keyring (if using keyring) and pass the "latest" password to `login`.

Reply:
On a related note, I'm checking the cookies involved in the login session and found a few things:
1. `LOGGED_IN_COOKIE` is actually unnecessary, not sure why I added this initially. Could have been used previously but now removed on their backend? Need further help on testing this but I've tested on `get_schedules` which requires the auth cookies, and removing it didn't make a difference.
2. There are three cookies returned on login: `__cfduid`, `tt_visitor_id` and `_mealpal_session`. The first two expire in 2020 (exactly 1 year from now), and the last one doesn't have an expiry. Whichever it is, it looks like we won't have to worry about cookies expiring. You can inspect `MealPal(...).cookies` after login to check the cookies, or use Chrome Dev tools.

If this is the case, I don't think we need to handle this issue of re-logins anymore? Simply preserve the cookies in the `MealPal` object as it is doing now. I doubt someone would run this program for a year straight.

Reply:
That's great news! I don't know anyone that would literally eat the same set of meals for an entire year lol. I would ballpark guess that 1 month spans would be reasonable for long-lived session, before user input can be expected again to reconfigure meal choices.

I'll update #6 with notes, we can discuss if we should revert the work done for #15 .

Reply:
I'm going to update `execute_reserve_meal` to be in charge of fetching the password and passing it into `login`, and we can handle saving the cookie for #6 in a separate PR.